riscv: Support -mstack-protector-guard=tls

keith-packard opened the pull request:
Add support for using a thread-local variable with a specified offset for holding the stack guard canary value.

Closes: #46685

keith-packard committed:
[RISCV][ISelLowering] Support -mstack-protector-guard=tls

Add support for using a thread-local variable with a specified offset
for holding the stack guard canary value.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[clang][RISCV] Add stack protector tests

Add tests ensuring that the driver correctly handles stack protector
guard options.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[CodeGen][RISCV] Add test for TLS stack canary

Make sure the code generated for a TLS-relative stack guard canary
references the right location before and after intervening function
code.

Signed-off-by: Keith Packard <keithp@keithp.com>

topperc commented:
> offsetfor

offset for

topperc commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3664,12 +3680,18 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
           << A->getOption().getName() << Value;
       return;
     }
+    if (EffectiveTriple.isRISCV() && (Offset <= -2048 || Offset >= 2048)) {
```
I don't know a lot about stack protector. Where does this range come from? Is this trying to be 12-bit signed immediate? If so it should exclude 2048.

arichardson replied:
Should probably have a comment and use IsInt<12>() instead of range checks.

keith-packard replied:
Yeah, this is used as an offset from the TLS base register (tp). GCC doesn't allow arbitrary offsets for this value, permitting only a value that can be encoded as an immediate value in the load instruction. I mirrored those limits (slightly incorrectly) in my patch. Given that llvm can support arbitrary values as the compiler is happy to use multiple instructions as required, I think it's probably best to just remove this test instead of introducing an artificial limit to match gcc's behavior?

topperc replied:
Ok. If the backend supports any value, we can drop this.

keith-packard replied:
Done, and new bits pushed.

arichardson commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3644,13 +3645,28 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
       return;
     }
+    if (EffectiveTriple.isRISCV()) {
+      if (Value != "tls" && Value != "global") {
```
Some of these could probably be simplified if we did something like

```
SmallVector<StringRef, 4> SupportedModes;
if (EffectiveTriple.isX86() || EffectiveTriple.isARM() ||         EffectiveTriple.isThumb() || EffectiveTriple.isRISCV()) {
    SupportedModes = {"global", "tls"};
} else if (EffectiveTriple.isAArch64()) {
    SupportedModes = {"global", "sysreg"};
}
if (!llvm::contains(SupportedModes, Value)) {
      D.Diag(diag::err_drv_invalid_value_with_suggestion)
          << A->getOption().getName() << Value << "sysreg global";
      return;
}
```

keith-packard replied:
Yeah, I went for the easiest-to-review mechanism that just created new RISC-V specific clauses instead of attempting to mix them into the existing ones. It's not exactly a performance-critical area of code.

arichardson replied:
Fair enough - this code code do we a bit of refactoring to remove duplication but that doesn't need to be part of this change.

keith-packard commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3644,13 +3645,28 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
+            << A->getOption().getName() << Value << "tls global";
+        return;
+      }
+      if (Value == "tls") {
```
Could probably be, but I felt it would be easier to review if I didn't attempt to refactor the code at the same time new functionality was being added.

topperc commented on llvm/test/CodeGen/RISCV/stack-guard-tls.ll:
```diff
@@ -0,0 +1,19 @@
+; RUN: llc -mtriple=riscv64-unknown-elf < %s | \
```
Use `llvm/utils/update_llc_test_checks.py` to generate the checks.

keith-packard replied:
Thanks. I wondered how that was supposed to work.

topperc commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3681,6 +3698,11 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
       return;
     }
+    if (EffectiveTriple.isRISCV() && Value != "tp") {
+      D.Diag(diag::err_drv_invalid_value_with_suggestion)
```
Is there a test for this error?

keith-packard replied:
Thanks, I missed that one. Added.

topperc commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3644,13 +3645,28 @@ static void RenderSSPOptions(const Driver &D, const ToolChain &TC,
       return;
     }
+    if (EffectiveTriple.isRISCV()) {
+      if (Value != "tls" && Value != "global") {
```
Is this also enabling "global" as a value? Is that supported in the backend?

keith-packard replied:
Yes, that's the existing non-TLS version which uses `__stack_chk_guard`. It's the default mode. I could add a test to make sure it still works?

keith-packard replied:
Added a test to make sure -mstack-protector-guard=global still works correctly after the patch. Also added positive tests to make sure valid driver arguments get passed through to the backend.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21228,6 +21228,14 @@ Value *RISCVTargetLowering::getIRStackGuard(IRBuilderBase &IRB) const {
   if (Subtarget.isTargetAndroid())
     return useTpOffset(IRB, -0x18);
 
+  Module *M = IRB.GetInsertBlock()->getParent()->getParent();
```
I think `getParent()->getParent()` can be replaced with `getModule()`

keith-packard replied:
Thanks, will review.

keith-packard replied:
Yup, getModule is just `return getParent()->getParent();`. Done.

topperc reviewed (approved):
LGTM

keith-packard committed:
[CodeGen][RISCV] Add test for global stack canary

Make sure the code generated for a global stack guard canary
references __stack_chk_guard before and after intervening function
code.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> LGTM

Thanks so much for all the review and help here; really appreciate it.

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,50 @@
+; RUN: llc -verify-machineinstrs < %s | \
+; RUN: FileCheck %s
+
+target triple = "riscv64-unknown-linux-gnu"
```
Convention is to put it in the RUN line

keith-packard replied:
sorry, copied this from llvm/test/CodeGen/AArch64/stack-guard-sysreg.ll which did it this way. I'll switch.

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,50 @@
+
+target triple = "riscv64-unknown-linux-gnu"
+
+define dso_local void @foo(i64 %t) local_unnamed_addr #0 {
```
Inline #0 for clarity? And are dso_local / local_unnamed_addr needed?

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,50 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -verify-machineinstrs < %s | \
+; RUN: FileCheck %s
```
The line above isn't (currently) long enough to warrant wrapping, but if you do you need to indent it like normal code so it's obviously a continuation line

keith-packard replied:
Thanks. After moving the triple into the RUN line, it's long enough that wrapping looks cleaner. I found a couple of examples using indentation on the continuation line to follow.

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,50 @@
+; CHECK-NEXT:  .LBB0_2:
+; CHECK-NEXT:    call __stack_chk_fail
+  %vla = alloca i32, i64 %t, align 4
+  call void @baz(ptr nonnull %vla)
```
Do you need the nonnull?

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,50 @@
+
+declare void @baz(ptr)
+
+attributes #0 = { sspstrong uwtable }
```
I doubt you actually need uwtable? Probably you want to nounwind the function instead unless you're really trying to test the unwind CFI.

keith-packard replied:
nope, it's been removed.

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-tls.ll:
Same comments as -global

keith-packard replied:
yup, all fixes mirrored over there. Thanks for your careful review!

jrtc27 commented on clang/test/CodeGen/stack-protector-guard.c:
```diff
@@ -23,3 +26,9 @@ void bar(int x) {
+// RISCV: [[ATTR1]] = !{i32 1, !"stack-protector-guard", !"tls"}
+// RISCV: [[ATTR2]] = !{i32 1, !"stack-protector-guard-reg", !"tp"}
+// RISCV: [[ATTR3]] = !{i32 1, !"stack-protector-guard-offset", i32 44}
+
```
```suggestion
```

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,47 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv64-unknown-elf -verify-machineinstrs < %s \
```
This was previously linux-gnu not elf. If it works fine with a generic triple like that then it is preferred, but we also prefer shortening such triples to just riscv64, as the rest is implied.

jrtc27 commented on llvm/test/CodeGen/RISCV/stack-guard-global.ll:
```diff
@@ -0,0 +1,47 @@
+; RUN: llc -mtriple=riscv64-unknown-elf -verify-machineinstrs < %s \
+; RUN:     | FileCheck %s
+
+define void @foo(i64 %t) sspstrong {
```
What about the nounwind? Did you decide against that, or just miss it?

keith-packard replied:
You were right, it doesn't need nounwind; we're just testing the stack protector code here.

jrtc27 replied:
No as in I would suggest adding it so that you don't get CFI being tested. uwtable was what was removed (which _forces_ unwind info to be emitted), but because this function calls `@baz`, which might throw an exception, it still ends up emitting unwind info, unless you add `nounwind` here to tell it that it can assume it's not needed.

keith-packard committed:
[RISCV][ISelLowering] Use getModule() instead of getParent()->getParent()

This is a simple clean-up motivated by review of changes adding TLS
support for the stack canary. Using getModule() is a shorter and
clearer expression of the desired operation.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I've retitled the commit messages to match llvm conventions and added a clean-up to replace another instance of `getParent()->getParent()` with `getModule()`.

topperc reviewed (approved):
LGTM. I assume you need someone to push it?

keith-packard commented:
> LGTM. I assume you need someone to push it?

Yes please! The PowerPC one is waiting in the wings :-)
